src/net/atlas/download.rs

```diff
@@ -0,0 +1,1025 @@
+    }
+}
+
+impl PartialOrd for AttachmentsInventoryRequest {
+    fn partial_cmp(&self, other: &AttachmentsInventoryRequest) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Requestable for AttachmentsInventoryRequest {
+    fn get_url(&self) -> &UrlString {
+        &self.url
+    }
+
+    fn get_request_type(&self, peer_host: PeerHost) -> HttpRequestType {
+        let mut pages_indexes = HashSet::new();
+        for page in self.pages.iter() {
+            pages_indexes.insert(*page);
+        }
+        HttpRequestType::GetAttachmentsInv(
+            HttpRequestMetadata::from_host(peer_host),
+            None,
+            pages_indexes,
+        )
+    }
+}
+
+impl std::fmt::Display for AttachmentsInventoryRequest {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let url = &**self.get_url();
+        write!(f, "<Request<AttachmentsInventory>: url={}>", url)
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct AttachmentRequest {
+    pub content_hash: Hash160,
+    pub sources: HashMap<UrlString, ReliabilityReport>,
+}
+
+impl AttachmentRequest {
+    pub fn get_most_reliable_source(&self) -> (&UrlString, &ReliabilityReport) {
+        self.sources
+            .iter()
+            .max_by_key(|(_, v)| v.score())
+            .expect("Atlas: trying to select an Url out of an empty set")
+    }
+}
+
+impl Hash for AttachmentRequest {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.content_hash.hash(state)
+    }
+}
+
+impl Ord for AttachmentRequest {
+    fn cmp(&self, other: &AttachmentRequest) -> Ordering {
+        other.sources.len().cmp(&self.sources.len()).then_with(|| {
+            let (_, report) = self.get_most_reliable_source();
+            let (_, other_report) = other.get_most_reliable_source();
+            report.cmp(&other_report)
+        })
+    }
+}
+
+impl PartialOrd for AttachmentRequest {
+    fn partial_cmp(&self, other: &AttachmentRequest) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Requestable for AttachmentRequest {
+    fn get_url(&self) -> &UrlString {
+        let (url, _) = self.get_most_reliable_source();
+        url
+    }
+
+    fn get_request_type(&self, peer_host: PeerHost) -> HttpRequestType {
+        HttpRequestType::GetAttachment(HttpRequestMetadata::from_host(peer_host), self.content_hash)
+    }
+}
+
+impl std::fmt::Display for AttachmentRequest {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let url = &**self.get_url();
+        write!(f, "<Request<Attachment>: url={}>", url)
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
+pub struct AttachmentsBatch {
+    pub block_height: u64,
+    pub consensus_hash: ConsensusHash,
+    pub block_header_hash: BlockHeaderHash,
+    pub attachments_instances: HashMap<QualifiedContractIdentifier, HashMap<(u32, u32), Hash160>>,
+    pub attachments_instances_count: u64,
```

Comment:
Isn't this duplicate information from `attachments_instances`?

Reply:
not really, attachments_instances_count is the sum of all the attachments_instances for all the contract_ids.

Reply:
That's what I mean -- `attachments_instances_count` is derived state.  You might want to replace it with a method that just returns this count, by iterating through `attachment_instances`.  It removes the possibility of creating an `AttachmentsBatch` where `attachments_instances_count` and `attachment_instances` are inconsistent.

Reply:
Addressed with https://github.com/blockstack/stacks-blockchain/pull/2013/commits/de4dbf3d5d6bf01bd2dd07369cf8a0f5d6d03412